File: python/0242_ValidAnagram.py
def isAnagram(s: str, t: str) -> bool:
    # Short plan (plain English):
    # 1) If lengths differ -> false.
    # 2) Count letters in s, subtract for t.
    # 3) If any count != 0 -> false; else true.
    # Time: O(n), Space: O(1) for 26 letters (or O(k) for alphabet size).
    if len(s) != len(t):
        return False
    count = [0] * 26
    base = ord('a')
    for ch in s:
        count[ord(ch) - base] += 1
    for ch in t:
        i = ord(ch) - base
        count[i] -= 1
        if count[i] < 0:
            return False
    return all(c == 0 for c in count)

File: java/0242_ValidAnagram.java
class Solution {
    // Short plan (plain English):
    // 1) If lengths differ -> false.
    // 2) Count letters for s, subtract for t.
    // 3) Verify all zero.
    // Time: O(n); Space: O(1) for 26 letters (O(k) for alphabet size).
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        int[] cnt = new int[26];
        for (int i=0; i<s.length(); i++) cnt[s.charAt(i) - 'a']++;
        for (int i = 0; i<t.length(); i++){
            if (--cnt[t.charAt(i) - 'a'] < 0) return false;
        }
        return true;
    }
}
